engine/errors.go

```diff
@@ -0,0 +1,51 @@
+package engine
+
+import "errors"
+
+var (
+	// ErrFeatureNotSupported is a common error that may be returned from optional
+	// Engine methods to indicate the engine implementation doesn't support the
+	// given feature.
+	//
+	// Note, all methods are allowed to return this error, some methods are
+	// required, and may not return this error.
+	//
+	// When the worker encounters this error from an optional method, it should
+	// workaround if possible, but most likely resolve the task as "exception"
+	// with reason "malformed-payload".
+	ErrFeatureNotSupported   = errors.New("Feature not support by current engine")
+	ErrResourceNotFound      = errors.New("The referenced resource wasn't found")
+	ErrSandboxTerminated     = errors.new("The Sandbox has terminated")
+	ErrSandboxAborted        = errors.new("Exection of sandbox was aborted")
+	ErrNonFatalInternalError = errors.new("Engine encountered a non-fatal internal error")
+	ErrContractViolation     = errors.new("Engine has detected a contract violation")
+	ErrEngineIsSingleton     = errors.New("Engine cannot run multiple sandboxes in parallel")
+	ErrEngineNotSupported    = errors.New("Engine is not available in the current configuration")
+	ErrEngineUnknown         = errors.New("Engine with the given doesn't exist")
+)
```

Comment:
These are valid points. One problem I've seen is this pattern:

```go
func XYZ(...) (<result type>, error) {
    ...
    ..., err := <some function call>
    if err != nil {
        return nil, err
    }
    ...
    return result, nil
}
```

Here we are essentially throwing an error up the stack, much like in java you'd just add a throws clause to a method, so that a particular error type just gets passed back up (note this is one example where java would do this more succinctly).

```java
public <return type> XYZ(...) throws <error type> {
    ... = <some function call>;
    return result;
}
```

In the java version, this is absolutely ok, because when you get the error back, you have a stack trace to where the error occurred within `<some function call>` so debugging is simpler. However, in the go version, you don't have a stack trace, just an error message, and you don't know where within the `XYZ` function or any function it called, however deeply nested, the error occurred. This is really nasty, and often makes debugging really difficult, without the stack trace. For this reason I tend to lean on putting more information into error messages, in case they are handled in this way. Of course, if you are in control of the code that calls the methods, you can avoid this pattern above. If others might be calling your functions, it is nice to add this information.

Note using `panic` and `recover` might help, but normally it is recommended in go that `panic` and `recover` can be used internally, but exposed APIs should not panic, rather return errors.

Maybe something like [go-errors](https://bugsnag.com/blog/go-errors) might help us here.